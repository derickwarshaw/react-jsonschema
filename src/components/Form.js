import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { cloneDeep, set } from 'lodash';
import SchemaField from './SchemaField';
import { getDefaultState } from '../utils';
import { addValueToState, deleteIndexFromState } from '../reducers';

class Form extends Component {
  constructor(props) {
    super(props);
    this.state = {};
    this.handleAddItem = this.handleAddItem.bind(this);
    this.handleDeleteItem = this.handleDeleteItem.bind(this);
    this.handleFieldChange = this.handleFieldChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  componentWillMount() {
    // Set the initial state
    this.setState({
      formData: getDefaultState(this.props.schema, this.props.formData),
    });
  }

  /**
   * Adds a new item
   *
   * @param {Object} e
   * @param {String} e.path
   * @param {Object} e.schema
   */
  handleAddItem(e) {
    const newState = addValueToState(this.state, {
      path: e.path,
      schema: e.schema,
    });
    this.setState(newState);
  }

  /**
   * Deletes an existing item
   *
   * @param {Object} e
   * @param {Number} e.index
   * @param {String} e.path
   */
  handleDeleteItem(e) {
    const newState = deleteIndexFromState(this.state, {
      index: e.index,
      path: e.path,
    });
    this.setState(newState);
  }

  /**
   * Handles a field change
   *
   * @param {Object} e
   * @param {String} e.path
   * @param {String|Number|Boolean} e.value
   */
  handleFieldChange(e) {
    const newState = cloneDeep(this.state);
    set(newState, e.path, e.value);
    this.setState(newState);
  }

  /**
   * Checks required fields and executes the onSubmit callback function
   *
   * @param {Event} e
   */
  handleSubmit(e) {
    e.preventDefault();
    const errors = [];
    // check required fields
    if (this.props.schema.required && this.props.schema.required.length) {
      this.props.schema.required.forEach(key => {
        // if the required fields are empty
        if (!this.state.formData[key]) {
          errors.push(key);
        }
      });
    }
    if (errors.length) {
      // execute the onError callback
      if (this.props.onError) {
        this.props.onError.call(this, errors);
      }
    }
    // execute the onSubmit callback
    if (this.props.onSubmit) {
      this.props.onSubmit.call(this, e);
    }
  }

  render() {
    const { children, schema } = this.props;
    const { formData } = this.state;
    return (
      <form onSubmit={this.handleSubmit}>
        <SchemaField
          schema={schema}
          formData={formData}
          onChange={this.handleFieldChange}
          onAddItem={this.handleAddItem}
          onDeleteItem={this.handleDeleteItem}
        />
        {children}
      </form>
    );
  }
}

Form.propTypes = {
  schema: PropTypes.object,
  formData: PropTypes.object,
  onError: PropTypes.func,
  onSubmit: PropTypes.func,
  children: PropTypes.any,
};

export default Form;
