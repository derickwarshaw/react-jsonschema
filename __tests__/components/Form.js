import React from 'react';
import renderer from 'react-test-renderer';
import types from '../../src/constants/types';
import Form from '../../src/components/Form';

it('should render a <Form /> component', () => {
  const schema = {
    type: types.OBJECT,
    properties: {
      firstName: {
        type: types.STRING,
      },
      lastName: {
        type: types.STRING,
      },
    },
  };
  const formData = {
    firstName: 'Vu',
    lastName: 'Tran',
  };
  const component = renderer.create(
    <Form
      schema={schema}
      formData={formData}
    />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
