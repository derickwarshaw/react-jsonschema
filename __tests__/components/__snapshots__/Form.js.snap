// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a <Form /> component 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    style={
      Object {
        "marginBottom": 15,
      }
    }
  >
    <div
      style={
        Object {
          "marginBottom": 15,
        }
      }
    >
      <div
        data-css-vbwgzr=""
      >
        <label
          data-css-1luuo4y=""
          htmlFor="formData.firstName"
        />
        <input
          data-css-14dfslc=""
          name="formData.firstName"
          onChange={[Function]}
          type="text"
          value="Vu"
        />
      </div>
    </div>
    <div
      style={
        Object {
          "marginBottom": 15,
        }
      }
    >
      <div
        data-css-vbwgzr=""
      >
        <label
          data-css-1luuo4y=""
          htmlFor="formData.lastName"
        />
        <input
          data-css-14dfslc=""
          name="formData.lastName"
          onChange={[Function]}
          type="text"
          value="Tran"
        />
      </div>
    </div>
  </div>
</form>
`;
